src/cls/lock/cls_lock.cc

```diff
@@ -132,27 +132,50 @@ static int write_lock(cls_method_context_t hctx, const string& name, const lock_
  * @return 0 on success, or -errno on failure
  */
 static int lock_obj(cls_method_context_t hctx,
-                    const string& name,
+                    const string& lock_name,
                     ClsLockType lock_type,
-                    utime_t duration,
+                    const utime_t& duration,
                     const string& description,
                     uint8_t flags,
                     const string& cookie,
-                    const string& tag)
+                    const string& tag,
+                    int32_t bid_amount,
+                    const utime_t& bid_duration)
 {
+  struct BidRecord {
+    int32_t amount;
+    utime_t expiration;
+  };
+
+  // three-level map, from object id, to lock name, to lock requester, to BidRecord
+  using ClientBidMap = std::unordered_map<entity_name_t /* lock requester */,BidRecord>;
+  using LockBidMap = std::unordered_map<std::string /* lock name */,ClientBidMap>;
+  using ObjectBidMap = std::unordered_map<string /* obj id */,LockBidMap>;
+
+  // this data does not have to persist between restarts, but it has
+  // to persist between calls; therefore let's keep it in memory
+  static std::mutex bids_mtx;
+  static ObjectBidMap bids;
```

Comment:
i'm concerned about the use of global state here. shouldn't these `bids` be specific to one rados object? if so, this map should be stored in the object's xattr with the rest of `lock_info_t`

Reply:
Iow, no one is (unreasonably) trying to "be clever."  Folks are trying to implement an efficient work-sharing mechanism that is, weakly, stateful.  I think @cbodley's suggestion to stash some specific data in ObjectContext was made on the assumption that it would be easy, and not intrusive.  I buy that that isn't the case, but statefulness of the type intended here is not overreach, it's just ordinary application of standard art.

Reply:
The librados interface is designed around objects and the osd is set up to maintain durable state around those objects; and to handle their consistency. Nothing in any interface is set up to handle in-memory state or for the object_class api to be stated Uk separate from what it writes out to objects. A "weakly stateful" Ceph interface doesn't exist.

Now, if this is actually a critical capability to develop, we can talk about that. But trying to shoehorn it in through an accident of the class implementation is indeed not tenable. There needs to be either a fundamentally different solution or a fundamentally different approach to getting it into the OSD, with a new system that has defined semantics and a proper interface for dealing with...all the things the osd does.

Reply:
That said, there is a mechanism for users of the osd to communicate between each other and send around state. It's watch-notify. Can this bid-and-lock algorithm be constructed that way, instead?

Reply:
thanks for the feedback!

> It's watch-notify. Can this bid-and-lock algorithm be constructed that way, instead?

good point, let's try to build a design around that. multisite already uses watch-notify to coordinate [bilog trimming](https://github.com/ceph/ceph/blob/8dae4aa/src/rgw/store/rados/rgw_trim_bilog.cc#L56-L239), so that could serve as an example

Reply:
I had a quick discussion with @mattbenjamin and it sounds like the crux of the issue is that you can't update a hypothetical persistent version of this state in the event that the lock operation fails.  Is that a reasonable characterization of the issue?

IIRC, object class writes can return payloads now.  We could get around this by modifying the cls_lock behavior to always succeed if the client sent a bid (or add a differently named method with that behavior) and instead indicate success via the returned payload.

Reply:
> IIRC, object class writes can return payloads now.

Correct. It looks like on the client side you have to set the flag OPERATION_RETURNVEC, and then just fill "outdata" in the class (same as for a read method), and the OSD checks it's less than osd_max_write_op_reply_len and passes it back as part of the result. This looks to be used by rgw/cls_fifo_legacy.cc and rgw/rgw_notify.cc, plus examples in cls_hello.